### PR TITLE
Document SAT API.

### PR DESCRIPTION
**Pull request by NlightNFotis**
Add some basic explanation of how the SAT api we are working with is implemented.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/solvers/README.md [hunk 1]
```diff
@@ -228,7 +228,44 @@ To be documented.
...
+or clauses the solver is operating with (with `no_variables`).
+3. Simulate gate operations on literals, applying Tseitin's
+transformation on them. Tseitin's transformation converts a
+propositional formula `F` into an equisatisfiable CNF formula
+that is linear in the size of `F`.
+
+This interface is then extended by the various solver interfaces
+which implemenet the interface by hooking into the solver
+related functions that implement the operations that they
+abstract. Solvers for which specific drivers exist include Minisat,
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/README.md]: I'd remove the "specific" here, it doesn't seem to do anything

File: src/solvers/README.md [hunk 2]
```diff
@@ -228,7 +228,44 @@ To be documented.
...
+2. Generating new [literals](\ref literalt) and adding them
+as variables to the formula and returning the number of variables
+or clauses the solver is operating with (with `no_variables`).
+3. Simulate gate operations on literals, applying Tseitin's
+transformation on them. Tseitin's transformation converts a
+propositional formula `F` into an equisatisfiable CNF formula
+that is linear in the size of `F`.
+
+This interface is then extended by the various solver interfaces
+which implemenet the interface by hooking into the solver
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/README.md]: implement

File: src/solvers/README.md [hunk 3]
```diff
@@ -228,7 +228,44 @@ To be documented.
...
+related functions that implement the operations that they
+abstract. Solvers for which specific drivers exist include Minisat,
+Minisat2, Chaff, Picosat, Glucose, Cadical, Booleforce and Lingeling.
+
+As an example of that, the Minisat 2 interface (in `satcheck_minisat2.h`)
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/README.md]: For example?

File: src/solvers/README.md [hunk 4]
```diff
@@ -228,7 +228,44 @@ To be documented.
...
+implements a method `prop_solve()` that hooks into Minisat 2's
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/README.md:253]: :pick: I'd rephrase that as "the interface of Minisat 2", but mostly because I'm personally uncomfortable with the usage of apostrophes especially with weird names like "Minisat 2".

File: src/solvers/README.md [hunk 5]
```diff
@@ -228,7 +228,44 @@ To be documented.
...
+interface by initialising the variable list Minisat 2 will use,
+and then invoking the actual solver on the formula and checking
+whether the solver could manage find a satisfying assignment or
+not.
+
+For more details on how the particular drivers work, please refer
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/README.md]: Not sure if the please here is really necessary

File: src/solvers/README.md [hunk 6]
```diff
@@ -228,7 +228,44 @@ To be documented.
...
+to them in their interface and implementation files, which follow
+the naming pattern `satcheck_x`, where `x` is the name of the
+solver.
+
+We also support any solver that can hook into `ipasir`. This is a
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/README.md]: I'm assuming [this](http://www.cs.utexas.edu/users/moore/acl2/manuals/current/manual/index-seo.php/IPASIR____IPASIR) ipasir? If so, please make the first occurrence of the word a link to that.

File: src/solvers/README.md [hunk 7]
```diff
@@ -228,7 +228,44 @@ To be documented.
 
 \subsection sat-solving-api-section SAT solving API
 
-To be documented.
+The basic SAT solver interface is in [cnf](\ref cnft). This inherits
+from the [propositional logic decision procedure wrapper](\ref propt).
+
+The interface supports the following operations by default:
+1. Boolean operations on literals (like `and`, `or`, `xor`), etc.
+These take as input two [literal](\ref literalt) and return as
```
Comment (tautschnig) [inline src/solvers/README.md]: literals

File: src/solvers/README.md [hunk 8]
```diff
@@ -228,7 +228,44 @@ To be documented.
...
+output another [literal](\ref literalt).
```
Comment (tautschnig) [inline src/solvers/README.md]: This should likely reference https://en.wikipedia.org/wiki/Tseytin_transformation or some more original description of it.

File: src/solvers/README.md [hunk 9]
```diff
@@ -228,7 +228,44 @@ To be documented.
...
+2. Generating new [literals](\ref literalt) and adding them
+as variables to the formula and returning the number of variables
```
Comment (tautschnig) [inline src/solvers/README.md]: "and adding their variables"

File: src/solvers/README.md [hunk 10]
```diff
@@ -228,7 +228,44 @@ To be documented.
...
+or clauses the solver is operating with (with `no_variables`).
+3. Simulate gate operations on literals, applying Tseitin's
+transformation on them. Tseitin's transformation converts a
```
Comment (tautschnig) [inline src/solvers/README.md]: Ah, I thought 1. was describing the same?

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: aee8fa32e342b473236e0bcb84319e4071bda78b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99823799

File: src/solvers/README.md (hunk 6 above)
Comment (NlightNFotis) [reply src/solvers/README.md]: Yes, fair enough.

File: src/solvers/README.md (hunk 10 above)
Comment (NlightNFotis) [reply src/solvers/README.md]: It's weird if I'm honest, I don't know how to describe it better. At least from the preexisting function documentation, the `gate_x` functions seem to be implementing tseitin's transformation, and the `literal_x` functions are just setup functions that end up calling equivalent the `gate_x` ones. Both are part of the interface of `cnft`, and I don't know which ones are to be favoured (they seem to be doing the same thing, but with a different interface).

I'm open to suggestions on how to rephrase it in a way that makes more sense.

**Comment by NlightNFotis** [comment]
@hannes-steffenhagen-diffblue @tautschnig Outstanding comments have been addressed. Can I get another review on this please?

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 46acaa1efff7f433e0d44609d20a46e9c8a0a4f3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99915855

File: src/solvers/README.md (hunk 10 above)
Comment (tautschnig) [reply src/solvers/README.md]: I believe the `gate_*` simply should not be part of the public interface, AFAIK they aren't actually used outside the class! They are concrete implementations of the `lx` functions, which the `propt` interface prescribes. Would you mind adding a commit to this PR that shifts the `protected` up to line 45 of `cnf.h`?

**Comment by NlightNFotis** [comment]
@tautschnig This is ready for another review now.

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 33572ba0a57b3e1423c0642303cfc4c0e4b8398b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100087744

**Comment by tautschnig** [comment]
@NlightNFotis This needs a rebase to pick up the Windows CI fix.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 1d3aa5ca26972db3e83b7ce0c29d679fa4223477).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100119505
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.
